kapp_platforms/src/macos/events_mac.rs

```diff
@@ -587,16 +589,16 @@ extern "C" fn set_marked_text(
     _selected_range: NSRange,
     _replacement_range: NSRange,
 ) {
+    // "Marked text" represents text that is not yet confirmed by the user.
+    // IME input candidates are provided through marked text.
     unsafe {
         let marked_text_ref: &mut *mut Object = this.get_mut_ivar("markedText");
         let _: () = msg_send![(*marked_text_ref), release];
         let marked_text: *mut Object = msg_send![class!(NSMutableAttributedString), alloc];
-        let has_attr = msg_send![string, isKindOfClass: class!(NSAttributedString)];
-        if has_attr {
-            let () = msg_send![marked_text, initWithAttributedString: string];
-        } else {
-            let () = msg_send![marked_text, initWithString: string];
-        };
+
+        let composition = get_str_helper(string).to_string();
+
+        self::submit_event(Event::IMEComposition { composition });
```

Comment:
I'm not sure if what you've got so far works correctly, but one thing I immediately noticed is you forgot to fire IMEEndComposition 😛

Reply:
Oops. I pushed a new commit adding the `IMEEndComposition` event.